### PR TITLE
♻️ Replace _types module by asgiref.typing

### PR DESCRIPTION
**Pull request by Kludex**
Following what was discussed on https://github.com/encode/uvicorn/pull/991 about adding `asgiref`.

This PR only does what the title suggests, and adapts the current code in the easiest way so it doesn't break and pass mypy.

### CODE REVIEW

File: uvicorn/lifespan/on.py [hunk 1]
```diff
@@ -31,7 +49,7 @@ async def startup(self) -> None:
         # Keep a hard reference to prevent garbage collection
         # See https://github.com/encode/uvicorn/pull/972
 
-        await self.receive_queue.put({"type": "lifespan.startup"})
+        await self.receive_queue.put(LifespanStartupEvent(type="lifespan.startup"))
```
Comment (Kludex) [inline uvicorn/lifespan/on.py]: Not doing this gives me: 
```bash
uvicorn/lifespan/on.py:52: error: Type of TypedDict is ambiguous, could be any of ("LifespanStartupEvent", "LifespanShutdownEvent")
uvicorn/lifespan/on.py:52: error: Argument 1 to "put" of "Queue" has incompatible type "Dict[str, str]"; expected "Union[LifespanStartupEvent, LifespanShutdownEvent]"
```
If you have suggestions, I'm open to change it :kissing:

Comment (euri10) [reply uvicorn/lifespan/on.py]: I think if you do something like 
```
lste : LifespanStartupEvent = {"type": "lifespan.startup"}
await self.receive_queue.put(lste)
```
it should be ok, though I have not tested it

**Review by euri10 — CHANGES_REQUESTED** [review]

File: uvicorn/lifespan/on.py (hunk 1 above)
Comment (Kludex) [reply uvicorn/lifespan/on.py]: Yeah, it should work. But I don't see any real improvement. Is there any?

Comment (euri10) [reply uvicorn/lifespan/on.py]: > Yeah, it should work. But I don't see any real improvement. Is there any?

well not having a mypy error is one in my mind, I'm no mypy pro but found the TypedDict experience quite frustrating and solved most of my issues by doing that, I'm not married to it though so if you have a better idea I'm all ears.

Comment (Kludex) [reply uvicorn/lifespan/on.py]: Well, the idea would be the current implementation on the PR, but it's the same to me. I've never done what I propose on this PR tho, I always do what you suggest, so I'm going to change the implementation.

Comment (Kludex) [reply uvicorn/lifespan/on.py]: Done

File: uvicorn/lifespan/on.py [hunk 2]
```diff
@@ -30,8 +48,8 @@ async def startup(self) -> None:
         main_lifespan_task = loop.create_task(self.main())  # noqa: F841
         # Keep a hard reference to prevent garbage collection
         # See https://github.com/encode/uvicorn/pull/972
-
-        await self.receive_queue.put({"type": "lifespan.startup"})
+        startup_event: LifespanStartupEvent {"type": "lifespan.startup"}
```
Comment (euri10) [inline uvicorn/lifespan/on.py]:
```suggestion
        startup_event: LifespanStartupEvent = {"type": "lifespan.startup"}
```

Comment (euri10) [reply uvicorn/lifespan/on.py]: should help ;)

Comment (Kludex) [reply uvicorn/lifespan/on.py]: haha :kissing:

**Review by euri10 — APPROVED** [review]
thanks !

**Pull request merged**